src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs

```diff
@@ -182,9 +180,7 @@ public ServiceController[] DependentServices
                             throw new Win32Exception(lastError);
 
                         // allocate the buffer
-                        IntPtr enumBuffer = Marshal.AllocHGlobal((IntPtr)bytesNeeded);
-
-                        try
+                        using (var enumBuffer = new SafeAllocHHandle(Marshal.AllocHGlobal((IntPtr)bytesNeeded)))
```

Comment:
The pattern where both alloc and free is done within same function and the thing does not escape anywhere is fine without safe handles. Safe handles just add overhead, without really improving safety.

Safe handles are worth it for the extra safety when the handle is object state (stored in a field). Safe handles prevent situation where some other thread frees it while it is being used.

Reply:
Reversed, but if not a SafeHandle a disposable holder seems like it might have helped avoid the leak in `Run()` in ServiceBase.